### PR TITLE
Celery add support for anon tasks

### LINKED ISSUE #1002: Using CeleryInstrumentor and linking Celery task trace back to requested trace
Hi,

I have a FastAPI frontend (with FastAPIInstrumentor set up) that sends tasks to a Celery backend using CeleryInstrumentor. However I having issues with linking the traces together to see an end to end request.

When I view the trace information in Jaeger, the trace appears as 2 separate services, an API service and a Celery service. There is no information linking the 2 together to show 1 full service trace.

However I am able to partially correct this. If I send some context information manually to the Celery task, I can then attach the Context to the span inside the worker. This works to a point, as while all child spans are linked back the API, the main CeleryInstrumentor parent span is not updated therefor is not linked.

Am I missing something that would link or at least update the CeleryInstrumentor parent span back to the API spans.

API send celery task:

```
TRACER = set_traceing(__name__) # sets TracerProvider(Resource)
CELERY = Celery("tasks",
                broker="redis://localhost:6379",
                backend="redis://localhost:6379"
                )
CELERY.conf.task_default_queue = "request-controller-2-default"

def send_service_request_to_celery(obj_in, celery_task: str) -> AsyncResult:
     with TRACER.start_as_current_span("send_service_request_to_celery") as span:
        tracer_context = span.get_span_context()
        tracer_context_obj = schema.ContextBase(trace_id= tracer_context.trace_id, span_id= tracer_context.span_id, is_remote=True)

        try:
            celery_out: AsyncResult = CELERY.send_task(
                name=celery_task,
                task_id=service_request_article.service_task_id,
                args=[service_request_article.dict(),
                tracer_context_obj.dict()],
            )
        
        except KombuError as error:
            message = (
                f"Failed to send service {obj_in.service_operation} request to request controller via redis."
                f" Service id: {obj_in.service_id}"
            )
```

celery task
```
from celery import Celery
from celery.signals import worker_process_init

from opentelemetry import trace, context
from opentelemetry.exporter.jaeger.thrift import JaegerExporter
from opentelemetry.instrumentation.celery import CeleryInstrumentor
from opentelemetry.instrumentation.requests import RequestsInstrumentor
from opentelemetry.trace import NonRecordingSpan, SpanContext, TraceFlags

from time import sleep
from typing import Union

from app.core.tracer import set_traceing

RequestsInstrumentor().instrument()

TRACER = set_traceing(__name__)
CELERY = Celery("tasks",
                broker="redis://localhost:6379",
                backend="redis://localhost:6379")
CELERY.conf.task_default_queue = "request-controller-2-default"

@worker_process_init.connect(weak=False)
def init_celery_tracing(*args, **kwargs):
   
    # Configure the tracer to export traces to Jaeger
    jaeger_exporter = JaegerExporter(
        agent_host_name="localhost",
        agent_port=6831,
    )
    span_processor = BatchSpanProcessor(jaeger_exporter)
    trace.get_tracer_provider().add_span_processor(span_processor)
    
    CeleryInstrumentor().instrument()

@CELERY.task(name="processing_task", bind=True, track_started=True)
def broadband_service_request(self, service_obj_in, tracer_context, **kwargs) -> Union[dict, str]:
    ctx = trace.set_span_in_context(NonRecordingSpan(SpanContext(
        trace_id=tracer_context['trace_id'],
        span_id=tracer_context['span_id'],
        is_remote=False,
        trace_flags=TraceFlags(0x01) )))
    context.attach(ctx)
    with TRACER.start_as_current_span("processing_task_span"):
        print(service_obj_in)
        sleep(1)
        with TRACER.start_as_current_span("some_work"):
            sleep(5)
        with TRACER.start_as_current_span("some_more_work"):
            sleep(5)

```
Search result
image
Unlinked CeleryInstrumentor Span
image
Linked Spans with Updated Child Context
image

### PR DESCRIPTION
**Pull request by goatsthatcode**
# Description

This changes the way spans are initialized for the celery instrumentation hooks and handles the case where Celery tasks are not available in the registry and called anonymously using the `signature()` or `send_task()`

Fixes #1002, #609 
possibly addresses #784 

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Ran changes against reproduced issue in repo linked in issue -> https://github.com/goatsthatcode/opentelemetry-instrumentation-celery-example
- [x] Added Unit Test to cover new conditional branch
- [x] Passed existing test suite

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [ ] Changelogs have been updated
- [ ] Unit tests have been added
- [ ] Documentation has been updated

### CODE REVIEW

**Review by srikanthccv — COMMENTED** [review]
@goatsthatcode Did you run the reproducible example provided in https://github.com/open-telemetry/opentelemetry-python-contrib/issues/609 and can confirm it is fixed?

File: instrumentation/opentelemetry-instrumentation-celery/src/opentelemetry/instrumentation/celery/utils.py [hunk 1]
```diff
@@ -132,6 +132,11 @@ def attach_span(task, task_id, span, is_publish=False):
     NOTE: We cannot test for this well yet, because we do not run a celery worker,
     and cannot run `task.apply_async()`
     """
+    if task is None:
```
Comment (blumamir) [inline instrumentation/opentelemetry-instrumentation-celery/src/opentelemetry/instrumentation/celery/utils.py:135]: I'll copy the comment I wrote on the original PR in your fork for the wide discussion:

> I think the effect of this return statement is that the span (along with activation) will not be recorded into the CTX_KEY, which means that a later invocation of the _trace_after_publish signal will not be able to retrieve_span thus will not end the span.
Am I missing something?

Didn't actually try to run it, so I might just not fully understand the details of what causes the "publish" span to be ended.

**Comment by goatsthatcode** [comment]
Hi guys, sorry for the radio silence here. I got really sick this last week so was offline for a bit. I am picking up the remaining pieces today to make sure this at least ready to be merged (unless any additional comments come up).  Sorry for the delay :D

**Comment by sanketmehta28** [comment]
Please resolves build errors.

**Comment by yousifd** [comment]
@goatsthatcode can you please resolve the build errors?

**Review by Vivalldi — CHANGES_REQUESTED** [review]
I _believe_ these will take care of the linting issues

**Comment by sanketmehta28** [comment]
There are still some lint and generate build issues. Please do resolve them

File: instrumentation/opentelemetry-instrumentation-celery/tests/test_tasks.py [hunk 2]
```diff
@@ -77,3 +77,46 @@ def test_task(self):
...
+            time.sleep(0.05)
+
+        spans = self.sorted_spans(self.memory_exporter.get_finished_spans())
+        self.assertEqual(len(spans), 1)
+
+        producer = spans
+
+        self.assertEqual(
+            producer.name, "apply_async/app.hidden_task"
+        )
```
[the remark this answers is not in the document]
Comment (rahuldimri) [reply instrumentation/opentelemetry-instrumentation-celery/tests/test_tasks.py:121]: Are we still working on this issue ?

**Comment by naormalca** [comment]
@goatsthatcode i'm very interested in this fix, are you still working on it? if not, could i take it from here?

**Comment by srikanthccv** [comment]
succeeded by #1407

**Pull request closed without merging**